[NNAPI EP] Remove/Refactor shaper inference calculation code

YUNQIUGUO opened the pull request:
**Description**: Describe your changes.
As title.

The purpose of this pr is to eliminate as much of repetitive shape inference code in nnapi ep shaper struct.

For composed ops:
-BatchNorm
-Reshape
-Squeeze
-BatchMatmul
still contains some shape calculation impl/logic.

Dynamic shape functions is not touched yet. 

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Clean up redundant code as cpu shape inference impl for NNAPI EP.

yuslepukhin commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -34,8 +34,22 @@
-  ORT_RETURN_IF_ERROR(shaper.Transpose(data_input, perm, output));
+
+  auto calculate_transpose_output_shape = [&](const std::string& data_input,
+                                              const std::string& output, const std::vector<int32_t>& perm) -> Status {
```
>const std::vector<int32_t>& perm [](http://example.com/codeflow?start=73&length=32)

Conding standards mandate use of gsl::span<const T> to make sure a wide variety of containers is accepted.

yuslepukhin replied:
> std::string& outpu [](http://example.com/codeflow?start=51&length=19)

std::string_view for both args

yuslepukhin commented:
>   std::vector<uint32_t> input_indices;

InlinedVector per conding standards

---
Refers to: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:28 in b5cbdb4. [](commit_id = b5cbdb4cb6ec103965675e8b784d6f73c86cc19d, deletion_comment = False)

yuslepukhin commented:
>                          const std::string& perm_input, const std::vector<int32_t>& perm,

gsl::span

---
Refers to: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:22 in b5cbdb4. [](commit_id = b5cbdb4cb6ec103965675e8b784d6f73c86cc19d, deletion_comment = False)

yuslepukhin commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -47,7 +61,42 @@
+                                              const std::string& output) -> Status {
+      const Shape& input_dimen = shaper[data_input];
+      uint32_t input_size = Product(input_dimen);
+      std::vector<uint32_t> output_dimen(shape_value.size());
```
>std::vector<uint32_t> outpu [](http://example.com/codeflow?start=6&length=27)

InlinedVector or TensorShapeVector per conding standards.
Do no resize the vector only to throw away the values. Use reserve() + push_back()

yuslepukhin commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -85,8 +134,24 @@
-  ORT_RETURN_IF_ERROR(shaper.Split(input, axis, outputs));
+  auto calculate_split_output_shape = [&](const std::string& input,
+                                          int32_t axis,
+                                          const std::vector<std::string>& outputs) -> Status {
```
>const std::vector<std::string>& output [](http://example.com/codeflow?start=42&length=38)

Ditto

yuslepukhin commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:
```diff
@@ -34,10 +34,28 @@
+  const auto* node_arg = graph_viewer.GetNodeArg(name);
+  const auto* shape_proto = node_arg->Shape();
+  const auto& shape_dims = shape_proto->dim();
+  std::vector<uint32_t> shape(shape_proto->dim_size());
```
>  std::vector [](http://example.com/codeflow?start=0&length=13)

IninedVector

yuslepukhin reviewed (changes requested):
:clock1:

YUNQIUGUO commented:
> 🕐

thanks for the review and suggestions!

regarding updating `std::string` -> `std::string_view` (yeah definitely better without copies) and using `InlinedVector` require updating a bunch of interface/functions within NNAPI code, will gradually update them.

also will review and see other similar places that can be improved/refactored in nnapi implementation.

edgchen1 commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:
```diff
@@ -76,7 +169,7 @@ static Status AddSqueezeOp(ModelBuilder& model_builder,
   const auto& operand_types(model_builder.GetOperandTypes());
 
-  const auto& input_shape(shaper[input]);
+  const auto& input_shape = GetShapeInfoFromNodeArg(model_builder, input);
```
```suggestion
  const auto input_shape = GetShapeInfoFromNodeArg(model_builder, input);
```

edgchen1 commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -235,7 +301,8 @@ Status BuildBatchMatMul(ModelBuilder& model_builder, const NodeUnit& node_unit)
                       b_transposed = model_builder.GetUniqueName(b + "/transposed");
-    ORT_RETURN_IF_ERROR(AddNnapiTranspose(model_builder, gemm_b_inputs.front(), b_new_perm, {0, 2, 1}, b_transposed));
+    std::vector<int32_t> perm_data{0, 2, 1};
+    ORT_RETURN_IF_ERROR(AddNnapiTranspose(model_builder, gemm_b_inputs.front(), b_new_perm, gsl::make_span(perm_data), b_transposed));
```
```suggestion
    ORT_RETURN_IF_ERROR(AddNnapiTranspose(model_builder, gemm_b_inputs.front(), b_new_perm, AsSpan<int32_t>({0, 2, 1}), b_transposed));
```
AsSpan is from core/common/span_utils.h. more info in the coding standards doc: https://github.com/microsoft/onnxruntime/blob/main/docs/Coding_Conventions_and_Standards.md

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/shaper.h:
```diff
@@ -95,58 +31,6 @@ class Shaper {
   void Clear();
 
```
Can we get rid of this? 

It's generally simpler to create a new instance of the class if instead of having a 'clear' or 'reset' method that has to be kept in sync with all the member variables.

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -34,8 +35,22 @@
+
+  auto calculate_transpose_output_shape = [&](const std::string& data_input,
+                                              const std::string& output, const gsl::span<const int32_t> perm) -> Status {
+    const Shape& input_dimen = shaper[data_input];
```
nit: This doesn't really need to be a lambda given it's only called once.

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:
```diff
@@ -47,26 +72,102 @@
+    auto min_shape = shape1IsBigger ? shape2 : shape1;
+    for (int i = (int)max_shape.size() - 1,
+             j = (int)min_shape.size() - 1;
+         i >= 0 && j >= 0;
```
nit: don't use c-style casts. gsl::narrow or gsl::narrow_cast

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:
```diff
@@ -47,26 +72,102 @@
+    const Shape& shape2 = shaper[input2_name];
+
+    // broadcasting support
+    bool shape1IsBigger = shape1.size() >= shape2.size();
```
Broadcasting seems like a fairly generic thing. Do we have an existing helper that does that? If not, can we split this out into a generic helper?

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:
```diff
@@ -723,14 +851,22 @@
+
+  const auto* shape_proto = input_node_arg->Shape();
+  ORT_RETURN_IF(shape_proto == nullptr, "shape_proto cannot be null for: ", input_node_arg->Name());
+
```
Wondering if these checks are worth the binary size cost.

If we're always looking up the NodeArg via an input/output name for a Node that should _always_ exist. Otherwise the Graph infrastructure that allowed the Node instance to be created is fundamentally broken.

Previously it looks like we didn't check that Shape() was nullptr here as we just got a value from shaper with no error path. Was that incorrect, or would we never get to this point if the shape was nullptr? Naively I would have expected the op support checker to require a shape given the lack of support for dynamic shapes in NNAPI.

YUNQIUGUO replied:
Right. Just checked actually in Baseopsupportchecker::HasSupportedInputsOutputs, it already checks shape_proto is not null, also in each individual Opsupportchecker, it calls GetShape() helper, also requires shape to present.

so will never get to this point if the shape is a nullptr. removed the redundant check here.

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:
```diff
@@ -680,15 +806,17 @@
+  ORT_RETURN_IF_ERROR(GetShapeInfoFromNodeArg(model_builder, output, output_shape));
+  const OperandType output_operand_type(operand_types.at(input).type, output_shape);
+  shaper.AddShape(output, output_shape);
 
```
What's the reason for adding the shape to Shaper? Can we not just retrieve that same info using GetShapeInfoFromNodeArg if needed in the future? Possibly the internals of Shaper could make that call as needed?

YUNQIUGUO replied:
> What's the reason for adding the shape to Shaper? Can we not just retrieve that same info using GetShapeInfoFromNodeArg if needed in the future?

Yes. that's right. 

but if I remove these lines for adding the shape to Shaper, it will fail bunch of UT test cases in execution stage with a ` Status Message: unordered_map::at: key not found` error status return.

I investigated a bit and found that, 
take SliceOpBuilder as an example(Slice does not require any shaper inference impl code previously, but it has this single step to add output_shaper into Shaper: 
https://github.com/microsoft/onnxruntime/blob/8483b9c6e3bb391525e5c8249bdc9ce0cc1a5522/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc#L2728-L2731
if remove this line here for slice, SliceOpTest will fail at https://github.com/microsoft/onnxruntime/blob/8483b9c6e3bb391525e5c8249bdc9ce0cc1a5522/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L375

which is further from https://github.com/microsoft/onnxruntime/blob/8483b9c6e3bb391525e5c8249bdc9ce0cc1a5522/onnxruntime/core/providers/nnapi/nnapi_builtin/model.cc#L64-L68
where we retrieve shape info using name from shaper class. 

so I guess to modify this may involve more changes and possibly may require remove the shaper struct completely in nnapi if I understand correctly. 

There may be some other ways to work around this tho, open to suggestions.

skottmckay replied:
I was thinking when you construct Shaper you can give it a GraphViewer instance. Shaper::operator[] can look for a shape in shape_map_ first (for newly created shapes that the EP added) and if not found return the shape from the NodeArg in the GraphViewer for `key`.

That provides the same functionality without having to copy shape info into shape_map_ that is already present in the graph. AddShape would only need to be called for modified/new shapes.

YUNQIUGUO replied:
yeah this makes sense. thanks for the clarification.

addressing this now, along with some other changes.

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.h:
```diff
@@ -25,9 +25,8 @@ Status AddScalarOperand(ModelBuilder& model_builder, std::vector<uint32_t>& inpu
-                         const std::string& output);
+                         const std::string& perm_input, const gsl::span<const int32_t> perm,
+                         const std::string& output, const Shape* output_shape);
 
```
Is this called from anywhere other than AddToModelBuilderImpl? If not the output_shape is never nullptr and doesn't need to be passed as a pointer here.

YUNQIUGUO replied:
I think this optional output_shape is put here mainly for reusing it for both `output shape provided` and `output shape needs to computed` cases... e.g. for AddNnapiReshape called in BuildBatchMatMul:

https://github.com/microsoft/onnxruntime/blob/ebff15d74394b8549a5af0f5f772073a928b5a3a/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc#L209

AddNnapiTranspose is just trying to do something similar here.

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -19,24 +21,44 @@
+    auto calculate_transpose_output_shape = [&](const std::string& data_input,
+                                                const std::string& output,
+                                                const gsl::span<const int32_t> perm) -> Status {
+      const Shape& input_dimen = shaper[data_input];
```
It would be good to check if the compiler can optimize away all the overhead of these single call lambas given it's only called locally. If not it's adding unnecessary binary size. Same question applies in lots of places. 

Whilst the lambda adds some context via its name, I think that could equally be done via a comment and a local block.

e.g. wouldn't this do the same without all the lambda overhead?

```c++
  auto& shaper(model_builder.GetShaper());

  // Calculate transpose output shape
  {
      const Shape& input_dimen = shaper[data_input];
      ORT_RETURN_IF_NOT(perm.size() == input_dimen.size(), "Invalid perm is given!");
      size_t size = input_dimen.size();
      Shape output_dimen(size);
      for (size_t i = 0; i < size; i++)
        output_dimen[i] = input_dimen[perm[i]];

      shaper.AddShape(output, output_dimen);
      output_shape = &shaper[output];
  }
```

YUNQIUGUO replied:
agree. replaced all the single call lambdas with a local block

edgchen1 commented on onnxruntime/core/providers/nnapi/nnapi_builtin/model.h:
```diff
@@ -123,7 +123,7 @@ class Model {
   std::unordered_set<std::string> scalar_outputs_;
 
-  Shaper shaper_;
+  std::optional<Shaper> shaper_;
```
actually, how about keeping the data member as `Shaper shaper_;` and updating Model to have a constructor like this?
`Model::Model(const GraphViewer& graph_viewer) : ..., shaper_{graph_viewer} {}`

then in `ModelBuilder::Prepare()`, create the model with `nnapi_model_ = std::make_unique<Model>(graph_viewer_);` and use nnapi_model_'s Shaper instance after that. maybe ModelBuilder doesn't even need its own Shaper

skottmckay replied:
Agree. 

The ownership of the Shaper needs to be clarified so this doesn't need to be `optional`. I agree with Ed that the Model should own it given it needs it for the longest time. I don't see why we need to create the model inside ModelBuilder::Prepare though given it's a call to the default ctor.

Based on that I would consider creating the model directly in NnapiExecutionProvider::Compile with the Model ctor taking a GraphViewer and creating Model::shaper_. Add a public getter for the Shaper to Model. ModelBuilder can use the Shaper instance from Model. Pass a Model& to the ModelBuilder ctor or ModelBuilder::Prepare. I'd prefer the former if we create the model prior to the ModelBuilder in Compile.

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.cc:
```diff
@@ -280,6 +280,28 @@ void GetFlattenOutputShape(const NodeUnit& node_unit, const Shape& input_shape,
+    // (checked already in BaseOpSupportChecker), call dim_value here only.
+    shape.push_back(SafeInt<uint32_t>(shape_dim.dim_value()));
+  }
+  // If we have an empty shape, (scalar input), we need to make it as {1} as
```
nit: shape_proto->dim() is iterable so could be

```c++
  for (const auto& shape_dim : shape_proto->dim()) {
    // shape_dim here can possibly have dim_param, but as dynamic shape is not supported in NNAPI for now
    // (checked already in BaseOpSupportChecker), call dim_value here only.
    shape.push_back(SafeInt<uint32_t>(shape_dim.dim_value()));
  }
```

skottmckay commented:
> 

Doesn't Model own the Shaper instance? Not sure why it needs a non-const Execution to get a Shaper vs. directly using its shaper_ member. 

Also, I would have expected Execution to have a reference to the Shaper in Model if needed instead of having one instance in Model and a separate instance in Execution.

---
Refers to: onnxruntime/core/providers/nnapi/nnapi_builtin/model.cc:69 in fb24eb7. [](commit_id = fb24eb70c5e176179a9adb64667dacd59b536fc9, deletion_comment = False)

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.h:
```diff
@@ -39,14 +40,14 @@ class ModelBuilder {
 
-  ModelBuilder(const GraphViewer& graph_viewer);
+  ModelBuilder(std::unique_ptr<Model> model, const GraphViewer& graph_viewer);
 
```
ModelBuilder doesn't need to take ownership here, so pass Model as a reference and not as a unique_ptr. 

Also Compile doesn't need to take a unique_ptr<Model> - it can just update the Model reference it was constructed with.

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/shaper.h:
```diff
@@ -17,138 +22,31 @@
+    }
+    const auto& shape = GetShapeInfoFromNodeArg(*graph_viewer_, key);
+    shape_map_.insert(std::make_pair(key, shape));
     return shape_map_.at(key);
```
why do we need to insert this locally?

YUNQIUGUO commented:
The latest branch actually fails UT test at this line of check here:

https://github.com/microsoft/onnxruntime/blob/d3b684cd9e36e085ffd4dced826e98cc50ef2bc3/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L309

with model outputs being some random number. still looking at this. but pushed the other changes for review first.

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc:
```diff
@@ -259,8 +259,8 @@ common::Status NnapiExecutionProvider::Compile(const std::vector<FusedNodeAndGra
-    ORT_RETURN_IF_ERROR(builder.Compile(nnapi_model));
+    ORT_RETURN_IF_ERROR(builder.Compile());
+    std::unique_ptr<nnapi::Model> nnapi_model(&builder.GetModel());
 
```
I think the original setup was fine here. The ModelBuilder internally creates the nnapi::Model and owns it until Compile is called, at which point the Compile implementation std::move's it into the arg.

Right now you have ModelBuilder owning the model, and when Compile is called you store a pointer to it in a unique_ptr meaning that unique_ptr will now delete the model, but so will the ModelBuilder dtor. 

The value of a unique_ptr should only ever be set via a std::move or an allocation.

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/model.h:
```diff
@@ -147,6 +148,8 @@
 
   void SetShaper(const Shaper& shaper) { shaper_ = shaper; }
 
```
This should go away. We don't want to allow our shaper to be replaced.

YUNQIUGUO commented on onnxruntime/core/providers/nnapi/nnapi_builtin/model.h:
```diff
@@ -165,12 +165,15 @@ class Execution {
   Execution& operator=(const Execution&) = delete;
 
-  const Shaper& GetShaper() const { return shaper_; }
+  // TODO: Before we validate if we actually need to keep a shaper instance for Execution (if we have dynamic shape
+  // outputs, shape can get updated during execution),
+  // we commented out Shaper here for now.
+  /* const Shaper& GetShaper() const { return shaper_; } */
```
Commented out the Shaper instance in Execution for now, due to the lack of support for updating dynamic dimensions in NNAPI (meaning that Execution temporarily is not required to keep a shaper instance for updating shapes.)

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:
```diff
@@ -38,37 +38,89 @@
+  ORT_RETURN_IF_ERROR(calculate_eltwise_output_shape(input1, input2, output1));
+  const OperandType output_operand_type(operand_types.at(input1).type, shaper[output1],
+                                        output_scale, output_zero_point);
+  ORT_RETURN_IF_ERROR(model_builder.AddOperation(ANEURALNETWORKS_MUL, input_indices,
```
Can you describe how/when you know to use the shaper instance to get a shape?

I see at two usage patterns but maybe there are more. 

One is here where we add something to the shaper in the lambda only to retrieve it immediately after. Is that an optional usage of Shaper and we could just provide a Shape& to the lambda or is there an implicit reason why we must call AddShape?

The other is in places like ConvOpBuilder::AddToModelBuilderImpl where we retrieve the shape of a weight from Shaper, but it's not clear to me when that was added to shaper or how I would know I'm meant to get it from there vs. the NodeArg. I assume these are cases where we've added an initializer with a different layout, but that's really opaque. 

The advantage of Shaper::operator[] calling GetShapeInfoFromNodeArg is that someone updating the NNAPI EP code can always use that to (hopefully) get a valid result. i.e. it abstracts away that a shape might come from values added with AddShape or from NodeArg shape info.

YUNQIUGUO replied:
> The advantage of Shaper::operator[] calling GetShapeInfoFromNodeArg is that someone updating the NNAPI EP code can always use that to (hopefully) get a valid result. i.e. it abstracts away that a shape might come from values added with AddShape or from NodeArg shape info.

I agree that the approach of using Shaper::operator[] to call GetShapeInfoFromNodeArg internally is much more neat and can make it a more consistent looking especially when it comes to cases like retrieving shape of weight initializer from shaper. but  I thought we think it's not a good idea to keep a dangling graph_viewer reference to Shaper, so I reverted the GetShapeInfoFromNodeArg() call in operator[] as it's not viable to call now. (maybe there's a work around for this?)

> Is that an optional usage of Shaper and we could just provide a Shape& to the lambda or is there an implicit reason why we must call AddShape?

For this particular case of BatchNorm op, Yes. don't really have to call AddShape() here. will modify it.

> Can you describe how/when you know to use the shaper instance to get a shape?

Generally speaking, where we used `shaper[]` to retrieve shape now are all special cases I think.

1.  Cases as you said where we used helper function to add initializer (to be specific weight) in different layouts. like `AddInitializerInNewLayOut()` - which further called https://github.com/microsoft/onnxruntime/blob/9edc9465f0258fa4cf83d7975905afa9ce1ec436/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc#L450-L453 that's where we add shape into shaper.
Similar things happened in another place inside GemmOpBuilder where we called `AddInitializerTransposed` for the second input for a case when bias not provided.

2. Cases involving composed operations where we want to save the shape for later. 
   - Most of the  shape retrieval happened in op_builder_helper.cc file 
   - Here for eltwise operation of BatchNorm op (potentially can get rid of)
   - One place in GemmOpBuilder where we added a SqueezeAll operation. (but for this, it is just a particular log output which if we don't mind can also removed the usage of shaper[].) https://github.com/microsoft/onnxruntime/blob/9edc9465f0258fa4cf83d7975905afa9ce1ec436/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc#L1787-L1794

Everywhere else now are using shape from NodeArg/can call GetShapeInfoFromNodeArg().

but yeah, it does not look ideal now as the way how we get shape now doesn't seem like a unified/general way.

skottmckay commented:
Can you address the cpplint warnings too please?

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -19,24 +21,42 @@ using android::nn::wrapper::OperandType, android::nn::wrapper::Type;
+    }
+    const auto shape = shaper[output];
+    output_shape = &shape;
+  }
```
You're setting a pointer to the address of a local variable which will almost immediately become invalid. 

If we need to support calling this with an empty output_shape (that seems unnecessary though - caller can ignore the output_shape) pass in std::optional<Shape>&. Otherwise pass in Shape& and assign it here (i.e. `output_shape = shaper[output];`

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -47,15 +67,48 @@ Status AddNnapiReshape(ModelBuilder& model_builder,
+    }
+    const auto shape = shaper[output];
+    output_shape = &shape;
   }
```
Ditto. Can't assign pointer to address of local variable.

skottmckay commented:
> using InitializerMap = std::unordered_map<std::string, const ONNX_NAMESPACE::TensorProto&>;

Can we use InlinedVector for Shape?

---
Refers to: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.h:28 in 29fd0d4. [](commit_id = 29fd0d431591671988ce884de6e3a27a14a6e17b, deletion_comment = False)

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -85,9 +138,22 @@ Status AddNnapiSplit(ModelBuilder& model_builder,
+  // Calculate split output shape
+  {
+    const auto& input_shape = shaper[input];
+    const auto count = gsl::narrow<int32_t>(outputs.size());
```
`const auto input_shape`

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -325,13 +395,35 @@ Status BuildBatchMatMul(ModelBuilder& model_builder, const NodeUnit& node_unit)
+        std::vector<Shape> dimens;
+        for (const auto& input_name : inputs) {
+          const Shape& dimen = shaper[input_name];
+          dimens.push_back(dimen);
```
`const Shape dimen`

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.h:
```diff
@@ -25,15 +25,14 @@ Status AddScalarOperand(ModelBuilder& model_builder, std::vector<uint32_t>& inpu
-                         const std::string& output);
+                         const std::string& perm_input, const gsl::span<const int32_t> perm,
+                         const std::string& output, std::optional<Shape> output_shape);
 
```
This needs to be a reference doesn't it for the caller to see the update?

skottmckay replied:
I think you can remove the output_shape arg from both AddNnapiTranspose and AddNnapiReshape.

The code in the implementation for when the input arg is nullptr matches the old behavior - namely it updates a value in Shaper with the transposed/reshaped value. I believe you could remove the arg, and always update the value in Shaper.

The calling code is either not passing in the output_arg, or it's not doing anything with it after the call. i.e. it doesn't care.

skottmckay commented:
> using InitializerMap = std::unordered_map<std::string, const ONNX_NAMESPACE::TensorProto&>;

NNAPI SetDimensions takes a vector so probably not. 

---
In reply to: [1246211836](https://github.com/microsoft/onnxruntime/pull/12618#issuecomment-1246211836) [](http://example.com/codeflow?ancestors=1246211836,1246211836)

---
Refers to: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.h:28 in 29fd0d4. [](commit_id = 29fd0d431591671988ce884de6e3a27a14a6e17b, deletion_comment = False)

YUNQIUGUO commented:
> > using InitializerMap = std::unordered_map<std::string, const ONNX_NAMESPACE::TensorProto&>;
> 
> NNAPI SetDimensions takes a vector so probably not.
> 
> In reply to: [1246211836](https://github.com/microsoft/onnxruntime/pull/12618#issuecomment-1246211836)
> 
> Refers to: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/helper.h:28 in [29fd0d4](https://github.com/microsoft/onnxruntime/commit/29fd0d431591671988ce884de6e3a27a14a6e17b). [](commit_id = [29fd0d4](https://github.com/microsoft/onnxruntime/commit/29fd0d431591671988ce884de6e3a27a14a6e17b), deletion_comment = False)

um I just updated Shaper::Shape to use InlinedVector<uint32_t> though. For SetDimensions also updated it to take Shape- InlinedVector<uint32_t>.

What's the reason why we can not update places like SetDimensions from std::vector<uint32_t> -> InlinedVector?

skottmckay commented:
> What's the reason why we can not update places like SetDimensions from std::vector<uint32_t> -> InlinedVector?

Sorry - I assumed SetDimensions was part of the NNAPI interface and we were stuck with that type. If we can used InlinedVector that's much better for shapes.

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -282,15 +348,18 @@
-      const OperandType output_operand_type(operand_types.at(a).type, shaper[output]);
+      const Shape& input1_dimen = shaper[a];
+      const Shape& input2_dimen = shaper[b_transposed];  // num_units, input_size
+      Shape output_dimen{input1_dimen[0], input2_dimen[0]};
```
nit: input1_dimen/input2_dimen shouldn't be references

skottmckay commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -19,43 +21,88 @@
+    shaper.AddShape(output, output_dimen);
+  }
+  const auto output_shape = shaper[output];
+
```
nit: not really necessary to retrieve a shape you calculated locally in output_dimen from shaper

skottmckay reviewed (approved):
:shipit:

YUNQIUGUO commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc:
```diff
@@ -353,4 +436,30 @@ Status BuildBatchMatMul(ModelBuilder& model_builder, const NodeUnit& node_unit)
+                        "dim2: ", std::to_string(dim_min_shape));
+    }
+
+    if (dim_max_shape == 0 || dim_min_shape == 0) {
```
thanks for all the comments! believe addressed most of them.

for this one, sorry - had one question, if updated to the above, how about the line `max_shape[i] = 0` ? or is it by default equals to 0?

edgchen1 replied:
sorry, I got that wrong. how about this:
```c++
if (dims_max_shape == 1) {
  max_shape[i] = dim_min_shape;
}
```

basically, at that point we either know both dim_max_shape and dim_min_shape are equal, or at least one of them is 1. we're building the output with max_shape, so just update it if it has the 1 (meaning dim_min_shape has the value we want, which could be 0).

YUNQIUGUO commented on onnxruntime/core/providers/nnapi/nnapi_builtin/builders/shaper.cc:
```diff
@@ -1,537 +1,21 @@
-Status Shaper::UpdateShape(const std::string& name, const Shape& new_shape) {
-  const Shape& old_shape = shape_map_.at(name);
+/* Status Shaper::UpdateShape(const std::string& name, const Shape& new_shape) {
+  const Shape& old_shape = (*this)[name];
```
removed the commented out code also. yeah we can always added back and updated in the future.

also got rid of the shaper.cc file totally as there's nothing needs to be hold up there now. hopefully it's fine and clearer.

yuslepukhin reviewed (approved):
:shipit:
